Dpp 1639 eas changes

carlbm opened the pull request.

carlbm committed:
Replaced logic for Provider Adjustments
No failing tests. Logic not fully covered

carlbm committed:
All acceptance tests passing
Tests for ProviderPaymentsGroup

carlbm committed:
Added tests for provider payments calculator
Added acceptance tests for no change and a deletion

Pauljgraham commented on src/PeriodEnd/ProviderAdjustments/SFA.DAS.Payments.Calc.ProviderAdjustments/Services/ProviderPaymentsCalculator.cs:
```diff
@@ -0,0 +1,64 @@
+
+namespace SFA.DAS.Payments.Calc.ProviderAdjustments.Services
+{
+    public interface ICalculateProviderPayments
```
We should have a standard approach on where we put interfaces.

carlbm replied:
I vote for in the same file 👍

Pauljgraham commented on src/PeriodEnd/ProviderAdjustments/SFA.DAS.Payments.Calc.ProviderAdjustments/Services/ProviderPaymentsCalculator.cs:
```diff
@@ -0,0 +1,64 @@
+
+    public class ProviderPaymentsCalculator : ICalculateProviderPayments
+    {
+        public List<PaymentEntity> CalculatePayments(
```
Could we change the name to CalculatePaymentsARefunds or something ....

carlbm replied:
Changed

Pauljgraham commented on src/PeriodEnd/ProviderAdjustments/SFA.DAS.Payments.Calc.ProviderAdjustments/Domain/Extensions/ProviderPaymentsGroupExtensions.cs:
```diff
@@ -0,0 +1,21 @@
+{
+    public static class ProviderPaymentsGroupExtensions
+    {
+        public static PaymentEntity MakePayment(this ProviderPaymentsGroup source, decimal amount)
```
MakePayment sounds like a processing function, would it be better to call it CreatePaymentEntity. 

Also does it need to be an extension method? Doesn't it belong to the ProviderPaymentsCalculator as a private method?

carlbm replied:
Changed

Pauljgraham commented on src/PeriodEnd/ProviderAdjustments/DeployProviderAdjustments/sql/dml/06 PeriodEnd.ProviderAdjustments.Populate.Reference.Current.dml.sql:
```diff
@@ -16,28 +10,17 @@ INSERT INTO [Reference].[ProviderAdjustmentsCurrent] (
+	P.PaymentName [PaymentTypeName],
+	V.PaymentValue [Amount]
+	
+FROM ${EAS_Deds.FQ}.dbo.EAS_Submission_Values V
```
If OpenQuery was introduced previously for performance reasons, why is it being removed?

carlbm replied:
There are no cross-server joins in this query where before there was a join against a table in transient

Pauljgraham commented on src/PeriodEnd/ProviderAdjustments/SFA.DAS.Payments.Calc.ProviderAdjustments.UnitTests/Sandbox.cs:
```diff
@@ -0,0 +1,33 @@
+{
+    [TestFixture]
+    public class Sandbox
+    {
```
Are these tests needed? As it's not testing our code

carlbm replied:
Interesting for future devs? Although it's not testing our code, our code uses what's being tested

Pauljgraham commented on src/PeriodEnd/ProviderAdjustments/SFA.DAS.Payments.Calc.ProviderAdjustments/Domain/ProviderPaymentsGroup.cs:
```diff
@@ -0,0 +1,67 @@
+    public class ProviderPaymentsGroup : IEquatable<ProviderPaymentsGroup>
+    {
+        public long Ukprn { get; }
+        public int PaymentType { get; }
```
Should this be a enum?

carlbm replied:
It's more data than an enum type, badly named maybe 😄

Pauljgraham commented on src/PeriodEnd/ProviderAdjustments/SFA.DAS.Payments.Calc.ProviderAdjustments/Services/ProviderPaymentsCalculator.cs:
```diff
@@ -0,0 +1,64 @@
+                }
+            }
+
+            foreach (var previousPaymentGroup in groupedPreviousPayments)
```
This second loop is doing a refund if the submission. Is it possible to break this into a function which returns the list of repayments

carlbm replied:
Changed

Pauljgraham commented on src/PeriodEnd/ProviderAdjustments/SFA.DAS.Payments.Calc.ProviderAdjustments/Services/ProviderPaymentsCalculator.cs:
```diff
@@ -0,0 +1,64 @@
+            var alreadyProcessedGroups = new HashSet<ProviderPaymentsGroup>();
+            var processedSubmissions = new HashSet<Guid>(
+                previousPayments.Select(x => x.SubmissionId)
+                    .Intersect(earnings.Select(y => y.SubmissionId)));
```
I'm not sure why we need to intersect here? Won't the new earnings submissions be different from the existing submissions

carlbm replied:
We want to process earnings without payments and payments without earnings. The intersect is the bit that we don't want to process

Pauljgraham commented on src/PeriodEnd/ProviderAdjustments/SFA.DAS.Payments.Calc.ProviderAdjustments/Services/ProviderPaymentsCalculator.cs:
```diff
@@ -0,0 +1,64 @@
+                .Where(x => !processedSubmissions.Contains(x.SubmissionId))
+                .ToLookup(x => new ProviderPaymentsGroup(x));
+
+            foreach (var earningGroup in groupedEarnings)
```
Could we put this into a function which creates a list of new payments

carlbm replied:
Changed

Pauljgraham commented on src/PeriodEnd/ProviderAdjustments/SFA.DAS.Payments.Calc.ProviderAdjustments/ProviderAdjustmentsProcessor.cs:
```diff
@@ -36,29 +42,27 @@ public virtual void Process()
-                }
-            }
-            else
+        private void PopulateCollectionPeriod(IEnumerable<PaymentEntity> payments)
```
Does this method belong to the CalculatePayments Class.

carlbm replied:
It's not really doing anything around calculations and is more infrastructure related

carlbm committed:
Changes based on tech review

Commit:
Merge remote-tracking branch 'origin/develop' into DPP-1639-eas-changes

carlbm committed:
Updated EAS to:
Deal with changes that don't change the submission id
Updated PK to allow for refunds when the earnings disappear

carlbm committed:
Merge branch 'DPP-1639-eas-changes' of https://github.com/SkillsFundingAgency/das-providerpayments into DPP-1639-eas-changes

carlbm committed:
Merge branch 'develop' into DPP-1639-eas-changes

carlbm committed:
Updated collection period name for EAS
